pkg/addons/ensure.go

```diff
@@ -124,18 +126,42 @@ func collectAddons(s *state.State) (addonsToDeploy []addonAction) {
 					name: resources.AddonCSIAwsEBS,
 				},
 			)
-		} else {
-			// since k8s 1.23 CSIMigrationAWS is turn by default and require installation of AWS EBS CSI driver even if
+		} else if greaterThan23.Check(k8sVersion) {
+			// since k8s 1.23 CSIMigrationAWS is turned on by default and requires installation of AWS EBS CSI driver even if
 			// we don't use external ccm (https://github.com/kubernetes/kubernetes/pull/106098)
-			k8sVersion := semver.MustParse(s.Cluster.Versions.Kubernetes)
-
-			if greaterThan23.Check(k8sVersion) {
-				addonsToDeploy = append(addonsToDeploy,
-					addonAction{
-						name: resources.AddonCSIAwsEBS,
-					},
-				)
-			}
+
+			addonsToDeploy = append(addonsToDeploy,
+				addonAction{
+					name: resources.AddonCSIAwsEBS,
+				},
+			)
+		}
+	}
+
+	// Special case for AZURE
+	if s.Cluster.CloudProvider.Azure != nil {
+		if s.Cluster.CloudProvider.External {
+			addonsToDeploy = append(addonsToDeploy,
+				addonAction{
+					name: resources.AddonCCMAzure,
+				},
+				addonAction{
+					name: resources.AddonCSIAzureDisk,
+				},
+				addonAction{
+					name: resources.AddonCSIAzureFile,
+				},
+			)
+		} else if greaterThan23.Check(k8sVersion) {
+			// since k8s 1.23 CSIMigrationAzureDisk is turned on by default and requires installation of AzureDisk CSI driver
+			// even if we don't use external ccm
+			// note: in 1.23 CSIMigrationAzureFile is still turn off by default
+
+			addonsToDeploy = append(addonsToDeploy,
```

Comment:
I think we can also deploy AzureFile CSI since we already deploy AzureDisk CSI.

Reply:
i am not sure `CSIMigrationAzureFile` is defaulted to false in 1.23,  we deploy the AzureFile drive without enabling the  flag:

- which implementation is used In-tree or CSI driver
- what happens to exist file volume when we upgrade from 1.22 to 1.23

I will check that

Reply:
done :)